common/source/docs/common-ie24-fuelcell.rst

```diff
@@ -0,0 +1,204 @@
+
+.. Note::
+    Monitoring of the fuel cell's internal error codes for pre-arm checks and failsafes are only done on the battery monitor instance allocated for the electrical telemetry.  The electrical battery monitor instance must therefore be set to trigger failsafes based on error codes.
+
+
+Failsafe Low Action Error Code Group
+++++++++++++++++++++++++++++++++++++
+
+.. raw:: html
+
+   <table border="1" class="docutils">
+   <tbody>
+
+   <tr>
+   <th>Error Code</th>
+   <th>Definition</th>
+   </tr>
+
+   <tr>
+   <td>31</td>
+   <td>Start Denied</th>
+   </tr>
+
+   <tr>
+   <td>30</td>
+   <td>Pressure Alert</th>
+   </tr>
+
+   <tr>
+   <td>21</td>
+   <td>Battery Low</th>
+   </tr>
+
+   <tr>
+   <td>20</td>
+   <td>Pressure Low</th>
+   </tr>
+
+   <tr>
+   <td>11</td>
+   <td>SPM Lost</th>
+   </tr>
+
+   <tr>
+   <td>10</td>
+   <td>Reduced Power</th>
+   </tr>
+
+   </tbody>
+   </table>
+
+
+Failsafe Critical Action Error Code Group
++++++++++++++++++++++++++++++++++++++++++
+
+.. raw:: html
+
+   <table border="1" class="docutils">
+   <tbody>
+
+   <tr>
+   <th>Error Code</th>
+   <th>Definition</th>
+   </tr>
+
+   <tr>
+   <td>40</td>
+   <td>Battery Critical</td>
+   </tr>
+
+   <tr>
+   <td>33</td>
+   <td>Pressure Critical</td>
+   </tr>
+
+   <tr>
+   <td>32</td>
+   <td>System Critical</td>
+   </tr>
+
+   </tbody>
+   </table>
+
+
+Example
++++++++
+This example has been provided to make it clear how the failsafe actions are set.
+
+The battery monitors have been configured so that electrical telemetry data is on battery monitor 1 and hydrogen tank telemetry is on battery monitor 2:
+
+- :ref:`BATT_MONITOR <BATT_MONITOR>` = 17 (electrical data)
+- :ref:`BATT2_MONITOR <BATT2_MONITOR>` = 18 (tank data)
+
+The failsafe actions that correspond to the error code groups listed above are then set on battery monitor 1 as that is the one associated with the electrical telemetry data.
+
+- :ref:`BATT_FS_LOW_ACT <BATT_FS_LOW_ACT>` = 2 (RTL)
+- :ref:`BATT_FS_CRT_ACT <BATT_FS_CRT_ACT>` = 1 (land)
+
+A tank level failsafe has been set on battery monitor 2 so that the vehicle will return to land when the fuel level gets below 20%.  This is done by setting:
+
+- :ref:`BATT2_FS_LOW_ACT <BATT2_FS_LOW_ACT>` = 2 (RTL)
+- :ref:`BATT2_CAPACITY <BATT2_CAPACITY>` = 100
+- :ref:`BATTX_LOW_MAH <BATT_LOW_MAH>` = 20
+
+As previously mentioned, normal voltage failsafes can be used on this fuel cell model.  Hence, as a final belt-and-braces approach a critical voltage failsafe has been set on the electrical generator monitor instance to initiate an immediate landing if voltage gets to 3.3 V per cell:
+
+- :ref:`BATT_CRT_VOLT <BATT_CRT_VOLT>` = 39.6
+
+
+Driver Not Healthy
+------------------
+
+If you see the `Generator: Driver Not Healthy` message in the GCS then ArduPilot has not received any data packets from the fuel cell for at least 5 seconds.  Check the following for common causes to this issue:
```

Comment:
as a side note, I think we should drop "Driver" from the error message.  For other sensors (like optical flow, etc) we just say the sensor is unhealthy not the driver that consumes the sensor's data.

Reply:
Fixed here.  Will PR AP code shortly.